keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,276 @@
+
+Generally, the system's overall shared bandwidth consists of a sum of shared
+quotas defined by all buckets and an auto-tunable extra shared bandwidth. In
+this so-called shared bandwidth, a request w/ higher priority can always be
+executed prior to the requests w/ lower priority. Which is, at worst case,
+when requests from top priority completely occupies the shared bandwidth,
+then requests from lower bands can only be executed under its reserved quota.
+And those "wild" request doesn't match any buckets will be rejected.
+
+##### Overview
+
+Within each priority band, we have a set of buckets matching the priority and
+a WRR request queue. Request that missed reserved quota will be queued into
+the WRR queue w/ the weight of its bucket. Note that reserving quota of
+buckets at priority N can also re-claim requests queued at N.
+
+All requests will be pre-processed before it's actually executed by 3 stages:
+
+0. We find a matched bucket by applying BucketBinding API.
+1. *Reserving*: We check if the bucket's reserved quota. If it's sufficient,
+the request will consume one reserved quota from the bucket and be executed
+immediately.
+2. *Prioritizing*: We classify and enqueue the request into a priority band.
+Requests at higher priority will __always__ be executed prior to requests at lower.
+3. *Distributing*: We poll requests from each priority band's queue by
+[Weighted Round Robin](https://en.wikipedia.org/wiki/Weighted_round_robin)
+algorithm. If the request's matched bucket's weight is higher, its request
+will have more chance to be executed.
+
+Each stage works separately and solves separate goals.
+
+```
+                    +-----------+  enqueued  +--------------+  pending  +--------------+ polled by WRR
+Inbound Request --> | Reserving | -------->  | Prioritizing | --------> | Distributing |  ------->  Executed
+                    +-----------+            +--------------+           +--------------+
+              sufficient |             insufficient |                          |
+          reserved quota |        STSB shared quota |                          |
+                         v                          v                          v
+                 Executed Immediately           Rejected                    Time-out
+```
+
+
+##### API Model and APIServer flags
+
+
+
+```go
+type Bucket struct {
+    ...
+
+    // ReservedQuota defines a max concurrency of requests can run under guarantee w/o competing.
+    ReservedQuota int
+
+    // SharedQuota defines a shared concurrency added to the system's shared bandwidth.
+    SharedQuota int
+
+    // Priority assigns the bucket into a certain priority band.
+    Priority PriorityBand // Ranged in 1...N
+
+    // Weight defines how much weight its matched requests have in the queue.
+    Weight int
+}
+
+type BucketBinding struct {
+    ...
+
+    // Rules defines how we bind request to a bucket.
+    Rules []BucketBindingRule // WIP
```

Comment:
Can you write out the rules? Those are completely unaddressed by my own draft (I haven't looked at Mike's yet).

Reply:
[https://github.com/yue9944882/inflight-rate-limit/blob/851051455e66481a4a1d185d8011535a4ecafc0d/types.go#L43-L46](https://github.com/yue9944882/inflight-rate-limit/blob/851051455e66481a4a1d185d8011535a4ecafc0d/types.go#L43-L46)

for one binding object, we can have several rules which are logically joint by "AND" operand. and for one `Field` ("namespace", "verb", "user.name", "user.groups", etc.) we can set several values which are joint by "OR" operand. 

an example binding matching requests made by (kevin __or__ john) __and__ namespaced (default __or__ kube-system):

```yaml
apiVersion: "flowcontrol.k8s.io/v1"
kind: BucketBinding
rules:    
- field: "user.name"
  values:
  - kevin
  - johnny
- field: "namespace"
  values:
  - default
  - kube-system
```

Reply:
OK, I think we definitely need to get this discussed in the KEP. I don't think we want to permit arbitrary json path expression evaluation before things are put in the queue. I'd expect it to consider only things that are easily and cheaply available: user, user group (?), api group/version/kind, subresource, namespace, name. I can even imagine discussing a source IP CIDR match (not sure if it's a good idea, but it is cheap to evaluate), or various facts about a client cert (e.g., is signed by some ca)--but I think user/group is probably more general and cheaper.

Reply:
> api group/version/kind, subresource, namespace, name.

we don't know the`kind` until we unpack the payload and that unpacking only happens when we [en|de]code in the REST logic. additionally, we don't know the name either if it's a CREATE requests or we're creating w/ a `generateName`

how about making this more structurelized e.g.?

```
apiVersion: "flowcontrol.k8s.io/v1"
kind: BucketBinding
rules: 
   user: 
       names: [A, B] // named A or B
       groups: [A, B]  // contains A or B
   requestAttribute:
       groups: [...]
       versions: [...]
       resources: [...]
       namespaces: [...]
```

Reply:
> OK, I think we definitely need to get this discussed in the KEP. I don't think we want to permit arbitrary json path expression evaluation before things are put in the queue. I'd expect it to consider only things that are easily and cheaply available: user, user group (?), api group/version/kind, subresource, namespace, name.

I think that was his explanation.  @yue9944882 just expand here.



> I can even imagine discussing a source IP CIDR match (not sure if it's a good idea, but it is cheap to evaluate)

Strongly against.  We've all seen servers relying on this crushed as incoming connections simply lie to get priority and ignore results.